### PR TITLE
artwork: fix album art retrieval for m4a files.

### PR DESCRIPTION
**Pull request by xinyazhang**
1. Add function to mp4ff for handling binary blobs.
	- Previously mp4ff assume all tags are C string, which is not true for
	  album arts.
2. Add mp4_extract_art to artwork plugin to utilize mp4ff.
3. Link mp4ff to artwork plugin, now libmp4ff is mandatory for artwork.

Additional notes for PR:
It seems mp4ff uses a mixed coding style so I kept the original tabs unchanged.

### CODE REVIEW

**Comment by xinyazhang** [comment]
This should close #1543

File: plugins/artwork/artwork.c [hunk 1]
```diff
@@ -1507,6 +1508,56 @@ apev2_extract_art (const char *fname, const char *outname) {
...
+    const char* image_blob;
+    uint32_t sz = mp4ff_meta_get_coverart (mp4, &image_blob);
+    if (!image_blob) {
+        ret = -1;
+        goto closemp4;
+    }
+    trace (stderr, "will write mp4 cover art (%u bytes) into %s\n", sz, outname);
+    write_file (outname, image_blob, sz);
+
+closemp4:
```
Comment (Alexey-Yakovenko) [inline plugins/artwork/artwork.c]: When a 20-line function has 2 gotos -- something is suspicious about it.

That's one concern.. another concern being that you're using stdio file API, while you should be using the deadbeef's VFS, which supports reading from archives, etc.

File: plugins/libmp4ff/mp4ff.h [hunk 2]
```diff
@@ -102,7 +102,17 @@ int mp4ff_meta_get_totaltracks(const mp4ff_t *f, char **value);
 int mp4ff_meta_get_totaldiscs(const mp4ff_t *f, char **value);
 int mp4ff_meta_get_compilation(const mp4ff_t *f, char **value);
 int mp4ff_meta_get_tempo(const mp4ff_t *f, char **value);
-int32_t mp4ff_meta_get_coverart(const mp4ff_t *f, char **value);
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4ff.h]: Instead of changing the existing public API, you should add a new API.

File: plugins/libmp4ff/mp4ff.h [hunk 3]
```diff
@@ -102,7 +102,17 @@ int mp4ff_meta_get_totaltracks(const mp4ff_t *f, char **value);
...
+
+/*
+ * CAVEAT:
+ * 	1. Cover art MUST be retrived by this function, to avoid strdup in
+ * 	mp4ff_meta_find_by_name
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4ff.h]: The above comment is kind pointless.

File: plugins/libmp4ff/mp4ff.h [hunk 4]
```diff
@@ -102,7 +102,17 @@ int mp4ff_meta_get_totaltracks(const mp4ff_t *f, char **value);
...
+ * 	2. This function now returns the size of the image data, instead of 0/-1
+ * 	values proposed by the original libmp4ff.
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4ff.h]: This comment should go as documentation to the new API, which you should add.

File: plugins/libmp4ff/mp4ff.h [hunk 5]
```diff
@@ -102,7 +102,17 @@ int mp4ff_meta_get_totaltracks(const mp4ff_t *f, char **value);
...
+ *
+ * Note, user should not modify the data pointed by *value
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4ff.h]: This comment is unnecessary, because `const char **` implies this.

File: plugins/libmp4ff/mp4meta.c [hunk 6]
```diff
@@ -36,36 +36,67 @@
 #include "mp4ffint.h"
 
 
-
-static int32_t mp4ff_tag_add_field(mp4ff_metadata_t *tags, const char *item, const char *value)
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: Why not keep this function, and add another one with a different name? This would have resulted in less code changes.

Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: It's hard to say for sure, but it looks like you renamed this function to `mp4ff_tag_add_string_field`, and reimplemented it using the new function mp4ff_tag_add_blob_field_move.

Why did you need to rename this one, if its behavior didn't really change?

File: plugins/libmp4ff/mp4meta.c [hunk 7]
```diff
@@ -36,36 +36,67 @@
...
+
+    if (!ret) {
+        free(itemcopy);
+        free(valuecopy);
+    }
+
+    return ret;
+}
+
+#if 0 /* Nobody are using it, but keep the code for safe */
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: Looks like you had to comment this out because of the function rename above?

This API is not used, I think, because writing mp4 tags is disabled (for now), but if I need to enable it, which I plan to do soon, it would appear broken?

File: plugins/libmp4ff/mp4meta.c [hunk 8]
```diff
@@ -36,36 +36,67 @@
...
+/*
+ * Postcondition:
+ * 	+ tags->tags are set to NULL if realloc failed
+ * 	+ *pitem and *pvalue are set to NULL if assignment successes
+ * 		- otherwise they are untouched.
+ * 	+ tags->tags[tags->count].item and .value are set to *pitem and *pvalue if
+ * 		assignment succcess
+ * 		- otherwise they are set to NULL
+ */
+static int32_t mp4ff_tag_add_blob_field_move(mp4ff_metadata_t *tags, char **pitem, char **pvalue, uint64_t size)
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: The interface of this function is very confusing, and undocumented, as well as the name and behavior.

I don't see any real need for having to pass `char **` pointers,

File: plugins/libmp4ff/mp4meta.c [hunk 9]
```diff
@@ -36,36 +36,67 @@
...
     }
+
+    *pvalue = NULL;
+    *pitem = NULL;
+
+    tags->count++;
+    return 1;
 }
 
+static int32_t mp4ff_tag_add_string_field(mp4ff_metadata_t *tags, const char *item, const char *value)
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: This should be renamed back to mp4ff_tag_add_field.

File: plugins/libmp4ff/mp4meta.c [hunk 10]
```diff
@@ -283,19 +316,33 @@ static int32_t mp4ff_parse_tag(mp4ff_t *f, const uint8_t parent_atom_type, const
...
+                    datasize = subsize-(header_size+8);
+                    free(data);
+                    data = malloc(datasize);
+                    if ((uint32_t)mp4ff_read_data(f, data, (uint32_t)datasize) != (uint32_t)datasize) {
+                        free(data);
+                        data = NULL;
+                        datasize = 0;
+                    }
+                } else {
+                    if (data) {free(data);data = NULL;}
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: please put `free(data)` and `data=NULL` on separate lines.

File: plugins/libmp4ff/mp4meta.c [hunk 11]
```diff
@@ -314,10 +361,21 @@ static int32_t mp4ff_parse_tag(mp4ff_t *f, const uint8_t parent_atom_type, const
 	{
 		if (!done)
 		{
-			if (name == NULL) mp4ff_set_metadata_name(f, parent_atom_type, &name);
-			if (name) mp4ff_tag_add_field(&(f->tags), name, data);
+            /* Some files tagged by iTunes 10.1 have a name atom with empty string. */
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: The block below seems to have its indentation messed-up

File: plugins/libmp4ff/mp4meta.c [hunk 12]
```diff
@@ -314,10 +361,21 @@ static int32_t mp4ff_parse_tag(mp4ff_t *f, const uint8_t parent_atom_type, const
...
+             *
+             * However, data is not used elsewhere (freed immediately), thus
+             * copy can be done by move.
+              */
+            if (name) {
+                //fprintf(stderr, "[mp4ff] Assign %s (atom id: %d) to %s\n", name, (int)parent_atom_type, data);
+                mp4ff_tag_add_blob_field_move(&(f->tags), &name, &data, datasize);
+            }
 		}
-
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: Please don't re-format the existing code, when it's not changing otherwise.

File: plugins/libmp4ff/mp4meta.c [hunk 13]
```diff
@@ -226,6 +258,7 @@ static int32_t mp4ff_parse_tag(mp4ff_t *f, const uint8_t parent_atom_type, const
     char * name = NULL;
 	char * data = NULL;
 	uint32_t done = 0;
+    uint64_t datasize = 0;
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: The `datasize` value gets calculated from uint32_t variables below, so there's no need to use 64 bit type.

File: plugins/libmp4ff/mp4meta.c [hunk 14]
```diff
@@ -283,19 +316,33 @@ static int32_t mp4ff_parse_tag(mp4ff_t *f, const uint8_t parent_atom_type, const
...
-				}
+                } else if (parent_atom_type == ATOM_COVER) {
+                    /*
+                      * Cover data is basically a binary blob, which may include \0 in the middle,
+                     * and it also doesn't need \0 at the end
+                     */
+                    datasize = subsize-(header_size+8);
+                    free(data);
+                    data = malloc(datasize);
+                    if ((uint32_t)mp4ff_read_data(f, data, (uint32_t)datasize) != (uint32_t)datasize) {
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: You would not need the casts, if you used uint32_t for the datasize.

File: plugins/libmp4ff/mp4meta.c [hunk 15]
```diff
@@ -459,9 +517,22 @@ int32_t mp4ff_meta_get_tempo(const mp4ff_t *f, char **value)
     return mp4ff_meta_find_by_name(f, "tempo", value);
 }
 
-int32_t mp4ff_meta_get_coverart(const mp4ff_t *f, char **value)
+uint64_t mp4ff_meta_get_coverart(const mp4ff_t *f, const char **value)
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4meta.c]: With the current code, uint64_t type is overkill.

File: plugins/libmp4ff/mp4ffint.h [hunk 16]
```diff
@@ -154,10 +154,12 @@ typedef struct
 
 
 /* metadata tag structure */
+/* we need the size field because the value may not be a blob instead of a C string */
 typedef struct
 {
     char *item;
     char *value;
+    uint64_t size;
```
Comment (Alexey-Yakovenko) [inline plugins/libmp4ff/mp4ffint.h]: This needs to be uint32_t, because none of the code in mp4ff supports atoms larger than 32 bits allow, and I think mp4 format itself doesn't support that either.

**Comment by Alexey-Yakovenko** [comment]
I wrote all of my concerns, but it's quite likely that I will have more questions after everything above gets fixed.

**Comment by Alexey-Yakovenko** [comment]
I looked more at the code, and the whole thing about the cover atom. It doesn't really belong to metadata.

I think the right thing would be to store the cover data as a separate thing in the `mp4ff_t`, instead of `mp4ff_metadata_t`.

And keep the data in mp4ff_metadata_t string-only.

If we start adding binary blobs in there -- that would complicate working with the data in the input plugins. We'd need some way to know whether it's a string or not.

Then we need to see what happens when there are multiple cover atoms in the file.

**Comment by Alexey-Yakovenko** [comment]
Another thing to remember: the cover blobs are often large, and take CPU and RAM to load. That's especially important when adding files to playlist, and playing them back -- we need to make sure that the covers DON'T load during insert/play.

Therefore, we should have a flag, which indicates whether we're loading cover, or not.

That might be tricky, but should be doable by changing how the `parse_atoms_int` function works, and/or making a new API that loads just the cover, but skips everything else.

**Comment by Alexey-Yakovenko** [comment]
I have added the code necessary for loading multiple cover images, storing them in a linked list, and accessing them. Please see c40b5f832b7b87cf6abe7f6e0688c1e129e55344

The next step would be to add the code to use this from artwork plugin.

File: plugins/libmp4ff/mp4meta.c (hunk 6 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4meta.c]: Yes you're right.
The point is, it looks like the original author assumes every tag field is a C string, which is not true esp. for cover arts, and this assumption is the reason why mp4ff can't handle embedded cover arts.

So here I renamed add_field to add_string_field to emphasize this assumption for tag fields.

File: plugins/libmp4ff/mp4meta.c (hunk 7 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4meta.c]: I commented out this one simply for code clean up, because it's a static function and no code are using it.

File: plugins/libmp4ff/mp4meta.c (hunk 8 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4meta.c]: It's not needed but good for semantics imo.
After a successful assignment the ownership of *pitem and *pvalue is transferred to tags, so it's better to reset the original point to avoid unintended sharing.

File: plugins/libmp4ff/mp4ff.h (hunk 2 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4ff.h]: The problem is the existing public API has no point to work.
We need the length of the embedded image file, and getting the length will require the API returns the length in someway.

Therefore either we got a broken unnamed one, or a fixed one with the same ABI but different semantics, or a fixed one with ABI changed.

In my view the last one is better because it enforces mp4ff users to change the code accordingly instead of leaving an unfixed bug.

File: plugins/libmp4ff/mp4meta.c (hunk 10 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4meta.c]: Err, I just left the original mp4ff code there (with tabs replaced by spaces)
However I'll fix this anyway.

File: plugins/libmp4ff/mp4meta.c (hunk 11 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4meta.c]: Sure, mp4ff uses tabs and spaces in a mixed way.
Maybe I should send another commit to replace all tabs with spaces before?

File: plugins/libmp4ff/mp4meta.c (hunk 13 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4meta.c]: No, datasize is calculated from subsize, which is uint64_t.

File: plugins/libmp4ff/mp4ffint.h (hunk 16 above)
Comment (xinyazhang) [inline plugins/libmp4ff/mp4ffint.h]: MP4 does support it and the code can be found in mp4ff_atom_read_header.
Although most of mp4ff is implemented as "just works" without full compliance with MP4 specs.

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang you shouldn't be fixing anything related to mp4ff in your PR (please read the latest comments first). I wrote a different code, which doesn't require changing the mp4meta APIs or data structures, with some explanation why it's done this way.

Now we only need to add the necessary code to artwork plugin

**Comment by Alexey-Yakovenko** [comment]
@xinyazhang I'm sorry, I was wrong about the 64 bit stuff - seems like mp4ff supports 64 bit offests correctly (needs to be tested though). However, I don't see any real need to use 64 bit for storing cover art file size.

**Comment by xinyazhang** [comment]
Err, I don't think I can use 329c434 directly.
The problem is the tags are parsed and loaded when user calls mp4ff_open_read(_metaonly), when we have not chance to call mp4ff_set_load_covers.

How about recording the offsets and lengths during open and add an API to get the covers later?

**Comment by Alexey-Yakovenko** [comment]
hmm. I'll have a look, and get back to you. missed this part somehow.

**Comment by Alexey-Yakovenko** [comment]
How about 27ce7482667a3e9035220537c8682095bcdcd319?
Now you can call mp4ff_open_read_coveronly.

For now, it reads metadata + covers, but can be improved later.

**Comment by xinyazhang** [comment]
Thanks, this should work.

**Comment by xinyazhang** [comment]
Updates pushed.

By the way, do you have plan to remove the "cover extraction" step for local files?
I think we can let the UI read the embedded cover data directly, instead of relaying though a temporary cache file.

File: plugins/artwork/artwork.c [hunk 17]
```diff
@@ -1507,6 +1508,56 @@ apev2_extract_art (const char *fname, const char *outname) {
...
+        .user_data = fp
+    };
+    mp4ff_t *mp4 = mp4ff_open_read_coveronly(&cb);
+    if (!mp4) {
+        deadbeef->fclose (fp);
+        return -1;
+    }
+
+    mp4ff_cover_art_t* art_list = mp4ff_cover_get (mp4);
+    // FIXME: What shall we do if we have multiple covers in the same file?
```
Comment (Alexey-Yakovenko) [inline plugins/artwork/artwork.c]: Right now we only support 1 cover per file. Please don't put comments like this in the code -- that's not the place for the questions.

**Comment by Alexey-Yakovenko** [comment]
The code looks ok, I'll merge, but seems it's failing the build. I don't know why. Did you merge origin/master before applying your changes?

**Comment by Alexey-Yakovenko** [comment]
> By the way, do you have plan to remove the "cover extraction" step for local files?

Yeah, I kinda have a "plan" to make a new cover art plugin with a better API, which could use the covers directly from local files/folders, without copying them. But I don't know when I'm gonna start working on this.

**Comment by Alexey-Yakovenko** [comment]
we should possibly add an extra file type check (mp4, m4a, m4b, etc) to avoid attempting loading album art from any files. attempting to load every file can be VERY slow.

**Comment by xinyazhang** [comment]
> Did you merge origin/master before applying your changes?
Yes I did rebase (otherwise I couldn't test locally).

> we should possibly add an extra file type check (mp4, m4a, m4b, etc) to avoid attempting loading album art from any files
It's not a big deal for performance, since mp4ff will failed to open instantly after the first unknown atom, but you're right, we should check this rather than offloading to mp4ff.

**Comment by Alexey-Yakovenko** [comment]
> Yes I did rebase (otherwise I couldn't test locally).

the issue was with the -I order in AM_CFLAGS.. It was using the system libmp4ff, and failing (obviously)

> It's not a big deal for performance, since mp4ff will failed to open instantly after the first unknown atom

yeah, but it will still open the file, and try reading data from it, causing memory allocations, then clean up, etc -- that's a lot of operations, and they all add up.
